cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -1038,6 +1043,71 @@ public void setServiceInstances(HttpRequest request, HttpResponder responder,
     }
   }
 
+  /*************************** Queues APIs **************************************/
+  @DELETE
+  @Path("/queues")
+  public synchronized void clearQueues(HttpRequest request, HttpResponder responder,
+                                       @PathParam("namespace-id") String namespaceId) {
+    // synchronized to avoid a potential race condition here:
+    // 1. the check for state returns that all flows are STOPPED
+    // 2. The API deletes queues because
+    // Between 1. and 2., a flow is started using the /namespaces/{namespace-id}/apps/{app-id}/flows/{flow-id}/start API
+    // Averting this race condition by synchronizing this method. The resource that needs to be locked here is
+    // runtimeService. This should work because the method that is used to start a flow - startStopProgram - is also
+    // synchronized on this.
+    // check if there are actively 'RUNNING' flows
+    try {
+      List<ProgramRecord> flows = listPrograms(Id.Namespace.from(namespaceId), ProgramType.FLOW, store);
+      for (ProgramRecord flow : flows) {
+        String appId = flow.getApp();
+        String flowId = flow.getId();
+        Id.Program programId = Id.Program.from(namespaceId, appId, flowId);
+        ProgramStatus status = getProgramStatus(programId, ProgramType.FLOW);
+        if (status.getStatus().equals("RUNNING")) {
```

Comment:
safer to check if it's not stopped. Otherwise it could be 'starting' and transition to 'running' during the clear